### ISSUE TITLE
Use --lint-spec in proposal

### ISSUE THREAD

**Issue by Jack-Works** [tc39/ecmarkup#429]
Is it possible to use --lint-spec in the proposal? I tried it in https://github.com/tc39/proposal-pattern-matching/pull/251 but it reports:

```
[2022-03-28T13:17:02.550Z] Error: spec.emu:39:19: Cannot find name: 'LineTerminator'.
[2022-03-28T13:17:02.552Z] Error: spec.emu:39:44: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.552Z] Error: spec.emu:39:63: Cannot find name: 'LineTerminator'.
[2022-03-28T13:17:02.553Z] Error: spec.emu:46:44: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.553Z] Error: spec.emu:46:63: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.553Z] Error: spec.emu:47:39: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.554Z] Error: spec.emu:48:16: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.554Z] Error: spec.emu:48:35: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.554Z] Error: spec.emu:49:21: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.555Z] Error: spec.emu:65:12: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.555Z] Error: spec.emu:66:12: Cannot find name: 'Expression'.
[2022-03-28T13:17:02.556Z] Error: spec.emu:69:7: Cannot find name: 'BindingIdentifier'.
[2022-03-28T13:17:02.556Z] Error: spec.emu:72:7: Cannot find name: 'NullLiteral'.
[2022-03-28T13:17:02.556Z] Error: spec.emu:73:7: Cannot find name: 'BooleanLiteral'.
[2022-03-28T13:17:02.557Z] Error: spec.emu:74:7: Cannot find name: 'NumericLiteral'.
[2022-03-28T13:17:02.557Z] Error: spec.emu:75:7: Cannot find name: 'StringLiteral'.
[2022-03-28T13:17:02.558Z] Error: spec.emu:76:7: Cannot find name: 'RegularExpressionLiteral'.
[2022-03-28T13:17:02.558Z] Error: spec.emu:77:7: Cannot find name: 'RegularExpressionLiteral'.
[2022-03-28T13:17:02.558Z] Error: spec.emu:78:7: Cannot find name: 'TemplateLiteral'.
[2022-03-28T13:17:02.559Z] Error: spec.emu:79:11: Cannot find name: 'NumericLiteral'.
[2022-03-28T13:17:02.559Z] Error: spec.emu:80:11: Cannot find name: 'NumericLiteral'.
[2022-03-28T13:17:02.559Z] Error: spec.emu:91:11: Cannot find name: 'Elision'.
[2022-03-28T13:17:02.559Z] Error: spec.emu:93:32: Cannot find name: 'Elision'.
[2022-03-28T13:17:02.560Z] Error: spec.emu:107:7: Cannot find name: 'Elision'.
[2022-03-28T13:17:02.560Z] Error: spec.emu:111:7: Cannot find name: 'PropertyName'.
[2022-03-28T13:17:02.731Z] Error: spec.emu:27:12: could not find a definition for nonterminal BindingInterpolation
[2022-03-28T13:17:02.732Z] Error: spec.emu:30:7: could not find a definition for nonterminal BindingInterpolation
```

And I have another question about the grammar.

When should I use `::` and when should I use `:`? I read the https://tc39.es/ecma262/multipage/notational-conventions.html#sec-grammar-notation but didn't learn it well. It looks like in most cases I only need `MyNewSyntax :` is that correct?

**Comment by bakkot** [comment]
> Is it possible to use --lint-spec in the proposal

It's currently intended for the actual spec; some of the rules, like those, assume that you have a full specification to work with, so that all referenced grammar productions are defined. It might happen to work for some proposals, but it's not a support use.

> When should I use `::` and when should I use `:`?

You should use `::` for anything in the main grammar, which defines productions in terms of tokens, and `:` for anything in the lower-level grammar, which defines tokens in terms of code points. (And there's also `:::` for the regex pattern grammar.) So yes, `:` is almost always the thing you want.

**Comment by Jack-Works** [comment]
> some of the rules, like those, assume that you have a full specification to work with, so that all referenced grammar productions are defined.

I have installed and added it to the CLI parameter, is it possible to read productions from the definition?

**Comment by Jack-Works** [comment]
> You should use :: for anything in the main grammar, which defines productions in terms of tokens, and : for anything in the lower-level grammar, which defines tokens in terms of code points. (And there's also ::: for the regex pattern grammar.) So yes, : is almost always the thing you want.

Sorry still don't get this, can you give an example of `::` and `:` definitions? thanks!

**Comment by bakkot** [comment]
Sorry, I had those backwards in one of the places; I've fixed my comment now. That's what I get for writing comments before I'm awake.

Anyway, an example is like, `DecimalLiteral` is (in part) `DecimalLiteral :: DecimalIntegerLiteral . DecimalDigits ExponentPart`. Because this is `::`, this is defined over the sequence of code points which make up source. So e.g. you can't have a space between `1` and `.` in `1.0`

By contrast, `MultiplicativeExpression` is (in part) `MultiplicativeExpression : MultiplicativeExpression MultiplicativeOperator ExponentiationExpression`. Because this is defined with `:`, it's a part of the grammar defined over the token stream produced by the `::` grammar. So you _can_ have a space between parts of these productions, as in `x * y` (because whitespace is generally skipped in the token stream).

**Comment by Jack-Works** [comment]
Oh! thanks! I understand the difference now

**Comment by bakkot** [comment]
Hopefully this is fixed in version 12.0.2. You'll need to wait for https://github.com/tc39/ecma262/pull/2718 to land and release [@tc39/ecma262-biblio](https://www.npmjs.com/package/@tc39/ecma262-biblio) version 2 (or build and check in your own biblio file from that PR) before you can use it, though.
